shiny/render/_dataframe.py

```diff
@@ -0,0 +1,320 @@
+
+        Returns
+        -------
+        :
+            An object suitable for being returned from a `@render.data_frame`-decorated
+            output function.
+
+        See Also
+        --------
+        ~shiny.ui.output_data_frame
+        ~shiny.render.data_frame
+        """
+        import pandas as pd
+
+        self.data: pd.DataFrame = cast(
+            pd.DataFrame,
+            cast_to_pandas(
+                data,
+                "The DataGrid() constructor didn't expect a 'data' argument of type",
+            ),
+        )
+
+        self.width = width
+        self.height = height
+        self.summary = summary
+        self.row_selection_mode = row_selection_mode
+
+    def to_payload(self) -> object:
+        res: dict[str, Any] = json.loads(
+            # {index: [index], columns: [columns], data: [values]}
+            self.data.to_json(  # pyright: ignore[reportUnknownMemberType]
+                None, orient="split"
+            )
+        )
+        res["options"] = dict(
+            width=self.width,
+            height=self.height,
+            summary=self.summary,
+            row_selection_mode=self.row_selection_mode,
+            style="grid",
+        )
+        return res
+
+
+class DataTable(AbstractTabularData):
+    def __init__(
+        self,
+        data: object,
+        *,
+        width: Union[str, float, None] = "fit-content",
+        height: Union[str, float, None] = "500px",
+        summary: Union[bool, str] = True,
+        row_selection_mode: Union[
+            Literal["none"], Literal["single"], Literal["multi-toggle"]
+        ] = "none",
+    ):
+        """
+        Holds the data and options for a ``shiny.render.data_frame`` output, for a
+        spreadsheet-like view.
+
+        Parameters
+        ----------
+        data
+            A pandas `DataFrame` object, or any object that has a `.to_pandas()` method
+            (e.g., a Polars data frame or Arrow table).
+        width
+            A width for the data grid to occupy, in CSS units. The default is `fit-content`,
+            which sets the grid's width according to its contents. Set this to `100%` to use
+            the maximum available horizontal space.
+        height
+            A _maximum_ amount of vertical space for the data grid to occupy, in CSS units.
+            If there are more rows than can fit in this space, the grid will scroll.
+        summary
+            If `True` (the default), shows a message like "Viewing rows 1 through 10 of 20"
+            below the grid when not all of the rows are being shown. If `False`, the message
+            is not displayed. You can also specify a string template to customize the
+            message, for example: `"Viendo filas {start} a {end} de {total}"`.
+        row_selection_mode
+            Use `"none"` to disable row selection, `"single"` to allow a single row to be
+            selected at a time, and `"multi-toggle"` to allow multiple rows to be selected
```

Comment:
`multi-toggle` is a bit clunky. Any reason not to go with `multiple`?

Also, I just want to make sure we don't paint ourselves into a corner with the parameter name `row_selection_mode`. If we add region selection in the future (like a spreadsheet), can we do that without too much trouble?

Maybe instead we should have a `selection_mode` parameter, with values like `row-single` and `row-multi`?

Reply:
I was imagining `multiple` would be the OS-native-like way I demoed earlier (click, Shift-Click, Ctrl-Shift-Click, etc. all working the way Finder or File Explorer or native select boxes do).

The main inherent difference between `multiple` and `multi-toggle` is that in the latter case, you can click on and click off each row independently. In the former, you can't select more than one row without a modifier key, and you also go back to having none selected without a modifier key. Maybe the `multi-toggle` is just better/more intuitive and I should forget about the other mode?

Reply:
Re: `row_selection_mode` vs. `selection_mode`, I was thinking that there would be these different selection modes, and multiple of these could be active at the same time:

* Row: off, single, multiple
* Column: off, single, multiple
* Cell: off, single, multiple
* Range: off, on

I can think of three possibilities:

1. Separate selection mode parameters for row, column, cell, range. The downside is lots of parameters.
2. One selection mode parameter that takes a string or list. The downside is the type system doesn't help you avoid invalid combinations (like `selection_mode = ["row-single", "row-multi"]`)
3. One selection mode parameter that takes a named tuple: `selection_mode = {"row": "single", "column": "multiple"}`

We could also decide that allowing multiple simultaneous selection modes is not worth the conceptual overhead and complexity.

Reply:
I've changed `multi-toggle` to `multiple`. It's still `row_selection_mode` though. If you feel strongly about this, we can discuss, otherwise I'm inclined to ship it like this.